File: advent_of_code/rust/aoc2015/src/bin/01.rs
#![allow(warnings, unused)]
use utils::*;

fn run(problem: i32, lines: &Vec<String>) -> i32 {
    let mut floor = 0;
    for (index, c) in lines[0].chars().enumerate() {
        floor += match c {
            ')' => -1,
            '(' => 1,
            _ => 0,
        };
        if problem == 2 && floor < 0 {
            return (index + 1) as i32;
        }
    }
    floor
}

fn main() {
    let input_lines = get_input_lines(file!());

    let example1 = multiline_lines("
(())
");

    run(1, &example1).should_be(0);
    run(1, &input_lines).debug("Star 1").should_be(138);

    let example2 = multiline_lines("
()())
");

    run(2, &example2).should_be(5);
    run(2, &input_lines).debug("Star 2").should_be(1771);

    aoc_exit();
}

File: advent_of_code/cpp/2015/01.cpp
#include "../advent_of_code.hpp"

int run(int problem, vector<string> &lines) {
  const string &line = lines[0];

  int floor = 0;
  for (int i = 0; i < sz(line); ++i) {
    const char c = line[i];

    if (c == ')') {
      --floor;
    } else if (c == '(') {
      ++floor;
    }

    if (problem == 2 && floor < 0) {
      return i + 1;
    }
  }

  return floor;
}

int main() {
  aocInit();

  auto inputLines = getInputLines();

  run(1, inputLines) | debug("Star 1") | eq(138);
  run(2, inputLines) | debug("Star 2") | eq(1771);

  return aocExit();
}
